modules/shell/example_test.go

```diff
@@ -0,0 +1,35 @@
+package shell_test
+
+import (
+	"testing"
+
+	"github.com/gruntwork-io/terratest/modules/shell"
+)
+
+func ExampleCommand() {
+	var t *testing.T
+
+	out, err := shell.RunCommandAndGetOutputE(t, shell.Command{
+		Command: "sh",
+		Args:    []string{"-c", `echo "hello world" && echo "this command has failed" >&2 && exit 1`},
+	})
+	t.Logf("Failure.\n\tOutput: %v\n\tError: %v", out, err)
+
+	if err != nil {
+		o, ok := err.(*shell.ErrWithCmdOutput)
+		if !ok {
+			t.Errorf("did not get correct type. got=%T", err)
+			return
```

Comment:
NIT: can combine these two lines into `t.Fatalf`

Reply:
Or even reduce the whole if-block to a `require.IsType` one-liner.

Reply:
I usually try to keep code as "obvious" as possible; only importing stdlibs if possible.
But I don't have a strong opinion on this, you can decide on what the example should use.

Would the require usage be:
```go
var errOut *shell.ErrWithCmdOutput
require.IsType(t, errOut, err)
```

Reply:
If this is an automated test (`example_test.go`), then importing that library is fine, as downstream users won't be affected.

Note also that, in general, we put example infra code (e.g., Terraform code) in the top-level `examples` folder and example test code into the top-level `test` folder. So this example probably makes more sense in `test`, which no one imports anyway.

Reply:
As noted above, it's not a test, it's an example.
The advantage of having an example like that is that is shows up in the GoDoc. 
For example (no pun intended) https://godoc.org/github.com/pkg/errors#Cause / https://github.com/pkg/errors/blob/master/example_test.go#L106

That's why I hesitate from importing the `require` package; although it would work, the standard type assertion is more obvious to the regular dev.

Moving it to a different package would not make sense, as then it doesn't show up in the GoDoc anymore.
However, the `test` package you have is "different" anyway, as it is actual integration-like tests. I don't want to add a test here, but documentation.

Reply:
OK, sorry for the long delay, just coming back to this now.

Understood on this being an example. So far, in Terratest, we've only used actual tests as examples, and not official Go-native examples. 

###  Advantages of using tests

1. Terratest is _meant_ to be used as part of an automated test suite with the `t *testing.T` arg passed in. So using real test functions ends up being a more accurate example and encourages following the right practices.
1. Automated tests can run in parallel. Examples cannot. The Terratest test suite is already getting quite slow, and I worry about it getting 10x slower if we have lots of examples.
1. Many Terratest functions are not for pure unit testing, and can only be used with lots of context from the outside world: e.g., Terraform code, Packer templates, AWS credentials, running K8S cluster, etc. This sort of thing doesn't work well in the world of Go examples, which I believe are meant for pure unit tests only.

### Advantages of examples

1. They show up in GoDoc, and can even be executed in Go playground right from the browser.

### Overall

I'm leaning towards stick with using real tests as our "examples," but am open to other opinions and trade-offs we should consider. @yorinasub17 Would be curious to hear your thoughts too!

Reply:
Well, Go examples actually never get executed (unless in the playground from the browser).
So (2.) does not really apply here.  

The function itself is already tested, and Examples are purely for documentation purposes, demonstrating, in this case, how to extract the output from the returned error. As this also doesn't really need any kind of setup / terraform files, I thought that an example may be a perfect fit.

(my opinion, I'm okay with doing it according to your standards 🙂)

Reply:
> Well, Go examples actually never get executed (unless in the playground from the browser).
> So (2.) does not really apply here.

Actually, they do! If you specify the `// Output:` comment block, Go runs examples as part of your automated test suite and compares `stdout` to that comment block.

Reply:
Oh you're right, and it actually fails because of the nil `*testing.T` 😅 

Means I could either move it to the example-dir, or remove the `// Output:` block.

Reply:
Yea, I worry any example code we have that isn't executed will eventually be broken... So another vote for sticking with real automated tests for our examples.

Reply:
True, although they still get compiled at least. 
I could rewrite the example to be:
```go
func ExampleCommand() {
        t := new(testing.T)

        out, err := shell.RunCommandAndGetOutputE(t, shell.Command{
                Command: "sh",
                Args:    []string{"-c", `echo "hello world" && echo "this command has failed" >&2 && exit 1`},
                Logger:  logger.Discard,
        })
        t.Logf("Failure.\n\tOutput: %v\n\tError: %v", out, err)

        if err != nil {
                o, ok := err.(*shell.ErrWithCmdOutput)
                if !ok {
                        t.Fatalf("did not get correct type. got=%T", err)
                }
                fmt.Printf("Outputs:\n\tStdout: %s\n\tStderr: %s\n", o.Output.Stdout(), o.Output.Stderr())
        }

        // Output:
        // Outputs:
        //      Stdout: hello world
        //      Stderr: this command has failed
}
```

The test works this way, and the documentation is correct too.

Reply:
That's certainly an improvement, but I think in net, using a proper automated test is still going to be the better fit for Terratest.

Reply:
I agree with @brikis98 that we should stick to our integration test suite for the example. While the example may be a fit here, the problem is that this sets a precedent in our code base. Going from 0 examples to 1 example is actually a big change, as it implies that we want examples.

This leads to questions everytime someone comes into the code base:

- Why does `RunCommandAndGetOutputE` have an example and the other functions do not?
- Would you accept contributions that add examples for all the other functions, like `aws` package? (this is where the execution time becomes a big problem)
- I'm adding new functionality. Should I add an example?

Unless we have the intention to go through and add examples for most or all of our functions, I feel that it is safer from a maintainability standpoint to avoid them altogether.

Reply:
Alright, thanks for the feedback!

I moved the example to an actual test and ensured that the output is correct.
I did not, however, move it to the `test` folder. The way it currently is, it's more of a unit test than an example. 

Should I also add a separate example? or move that unit test to the example folder?

Reply:
I think this is fine as-is, thanks!